File: CompetitionSpring/src/main/java/soprajc/CompetitionSpring/model/Compte.java
package soprajc.CompetitionSpring.model;

import java.util.Arrays;
import java.util.Collection;
import java.util.List;
import java.util.Objects;

import javax.persistence.Column;
import javax.persistence.Embedded;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.Inheritance;
import javax.persistence.InheritanceType;
import javax.persistence.OneToMany;
import javax.persistence.Version;
import javax.validation.constraints.NotEmpty;

import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;

import com.fasterxml.jackson.annotation.JsonView;


@Entity
@Inheritance(strategy = InheritanceType.TABLE_PER_CLASS)
public abstract class Compte implements UserDetails {

	@Id
	@GeneratedValue(strategy = GenerationType.TABLE)
	@JsonView(JsonViews.Common.class)
	protected Integer id;
	
	@NotEmpty
	@JsonView(JsonViews.Common.class)
	protected String nom;
	
	@NotEmpty
	@JsonView(JsonViews.Common.class)
	protected String prenom;
	
	@NotEmpty
	@Column(unique=true)
	@JsonView(JsonViews.Common.class)
	protected String mail;
	
	@NotEmpty
	protected String password;
	
	@Embedded
	@JsonView(JsonViews.Common.class)
	protected Adresse adresse;
	
	@OneToMany(mappedBy = "compte")
	@JsonView(JsonViews.CompteWithReservation.class)
	protected List<Reservation> reservations;
	
	@Version
	protected int version;
	
	public Compte() {
	}
	
	public Compte(Integer id, String nom, String prenom, String mail, String password, String numero, String voie, String ville, String cp) {
		this.id = id;
		this.nom = nom;
		this.prenom = prenom;
		this.mail = mail;
		this.password = password;
		this.adresse = new Adresse(numero,voie,ville,cp);
	}

	public Compte(String nom, String prenom, String mail, String password, String numero, String voie, String ville, String cp) {
		this.nom = nom;
		this.prenom = prenom;
		this.mail = mail;
		this.password = password;
		this.adresse = new Adresse(numero,voie,ville,cp);
	}
	
	public Integer getId() {
		return id;
	}

	public void setId(Integer id) {
		this.id = id;
	}

	public String getNom() {
		return nom;
	}
	public void setNom(String nom) {
		this.nom = nom;
	}
	public String getPrenom() {
		return prenom;
	}
	public void setPrenom(String prenom) {
		this.prenom = prenom;
	}
	public String getMail() {
		return mail;
	}
	public void setMail(String mail) {
		this.mail = mail;
	}
	public String getPassword() {
		return password;
	}
	public void setPassword(String password) {
		this.password = password;
	}
	public Adresse getAdresse() {
		return adresse;
	}
	public void setAdresse(Adresse adresse) {
		this.adresse = adresse;
	}

	public List<Reservation> getReservations() {
		return reservations;
	}

	public void setReservations(List<Reservation> reservations) {
		this.reservations = reservations;
	}
	
	public int getVersion() {
		return version;
	}

	public void setVersion(int version) {
		this.version = version;
	}

	@Override
	public String toString() {
		return "Client [id=" + id + ", nom=" + nom + ", prenom=" + prenom + ", mail=" + mail + ", password=" + password
				+ ", adresse=" + adresse + "]";
	}

	@Override
	public int hashCode() {
		return Objects.hash(id);
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (getClass() != obj.getClass())
			return false;
		Compte other = (Compte) obj;
		return Objects.equals(id, other.id);
	}
	
	@Override
	public Collection<? extends GrantedAuthority> getAuthorities() {
		GrantedAuthority authority = null;
		if (getClass().getSimpleName().equals("Athlete")) {
			authority = new SimpleGrantedAuthority("ROLE_ATHLETE");
		}	else if (getClass().getSimpleName().equals("Journaliste")) {
			authority = new SimpleGrantedAuthority("ROLE_JOURNALISTE");
		}	else if (getClass().getSimpleName().equals("Organisateur")) {
			authority = new SimpleGrantedAuthority("ROLE_ORGANISATEUR");
		}	else if (getClass().getSimpleName().equals("Spectateur")) {
			authority = new SimpleGrantedAuthority("ROLE_SPECTATEUR");
		}
		return Arrays.asList(authority);
	}
	
	@Override
	public String getUsername() {
		return mail;
	}
	
	@Override
	public boolean isAccountNonExpired() {
		return true;
	}
	
	@Override
	public boolean isAccountNonLocked() {
		return true;
	}
	
	@Override
	public boolean isCredentialsNonExpired() {
		return true;
	}
	
	@Override
	public boolean isEnabled() {
		return true;
	}
}

File: CompetitionAngular/src/app/model/compte.ts
import { Adresse } from './adresse';
export class Compte {
  public constructor(
    private _id?: number | undefined,
    private _nom?: string | undefined,
    private _prenom?: string | undefined,
    private _mail?: string | undefined,
    private _dateNaissance?: Date | undefined,
    private _entreprise?: string | undefined,
    private _raisonSoc?: string | undefined,
    private _adresse?: Adresse | undefined
  ) {}

  public get id(): number | undefined {
    return this._id;
  }

  public set id(value: number | undefined) {
    this._id = value;
  }

  public get nom(): string | undefined {
    return this._nom;
  }

  public set nom(value: string | undefined) {
    this._nom = value;
  }

  public get prenom(): string | undefined {
    return this._prenom;
  }

  public set prenom(value: string | undefined) {
    this._prenom = value;
  }

  public get mail(): string | undefined {
    return this._mail;
  }

  public set mail(value: string | undefined) {
    this._mail = value;
  }

  public get dateNaissance(): Date | undefined {
    return this._dateNaissance;
  }

  public set dateNaissance(value: Date | undefined) {
    this._dateNaissance = value;
  }

  public get entreprise(): string | undefined {
    return this._entreprise;
  }

  public set entreprise(value: string | undefined) {
    this._entreprise = value;
  }

  public get raisonSoc(): string | undefined {
    return this._raisonSoc;
  }

  public set raisonSoc(value: string | undefined) {
    this._raisonSoc = value;
  }

  public get adresse(): Adresse | undefined {
    return this._adresse;
  }

  public set adresse(value: Adresse | undefined) {
    this._adresse = value;
  }
}
